conjure-runtime/src/client_factory.rs

```diff
@@ -301,16 +303,27 @@ impl ClientFactory {
             let service = service.to_string();
             move |c| c.merged_service(&service).unwrap_or_default()
         });
-
         let state_builder = self.state_builder(service);
+        Self::raw_client(state_builder, service_config, None)
+    }
 
-        let state = state_builder.build(&service_config.get(), None)?;
+    fn raw_client<T>(
+        state_builder: T,
+        service_config: Refreshable<ServiceConfig, Error>,
+        override_host_index: Option<usize>,
+    ) -> Result<Client, Error>
+    where
+        T: Borrow<StateBuilder> + 'static + Sync + Send,
```

Comment:
I'm trying to understand how this `Borrow` type works (and why it's useful vs. just using `&T`).

I read about it here https://doc.rust-lang.org/std/borrow/trait.Borrow.html. Is it sort of like `String`, in that you can just pass `T` directly, but it can be implicitly cast to `&T` (or `&str`) because it implements `Borrow`?

Reply:
It has some extra requirements that are used in some other places, but the key thing in this context is that it's a trait implemented by both `T` and `Arc<T>` that lets you get a `&T`. This function is called in 2 places. The per-host stuff needs to store the state builder in an arc, while the normal stuff up in client_inner doesn't and this bound makes both work.